Replace test images with images from Getty's Open Content Program.

MrBago opened the pull request.

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeatureizer.scala:
```diff
@@ -0,0 +1,177 @@
+/*
```
Nit: Rename this file from "DeepImageFeatureizer.scala" -> "DeepImageFeaturizer.scala"

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeatureizer.scala:
```diff
@@ -0,0 +1,177 @@
+  private val INPUT_BUFFER_COL = "__sparkdl_imageBuffer"
+
+  override def transformSchema(schema: StructType): StructType = {
+    schema.add(getOutputCol, VectorType)
```
Should we perform schema validation here? E.g. check the datatypes of `inputCol` & `outputCol` to make sure they're both strings? Here's an example of how it's done in Spark: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala#L151.

smurching replied:
Follow-up: If we do add schema validation, we should unit test it.

smurching commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
```diff
@@ -0,0 +1,80 @@
+    // check that we can materialize a row, and the type is Vector.
+    val vect = d1.select(col(outputColName)).first().getAs[Vector](0)
+
+    // Test that we can keep columns with names that match the output graph node name. test_net's
```
I'm a little confused by this test; how should DeepImageFeaturizer behave when its output column is already present in the DataFrame passed to transform? AFAICT the transformer currently overwrites existing columns if needed. See the docs for Dataset.withColumn: https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Dataset

smurching replied:
My recommendation would be to fail if an output column already exists, which would match the behavior in existing Spark transformers (`StringIndexer`, `Bucketizer`, `HashingTF`, etc)

MrBago replied:
Sorry this was leftover from when I was trying to do something very unnecessary. I've added better errors and a better test.

smurching reviewed (commented):
A few small comments -- besides nits, I had a few Qs that we can discuss in person if you'd like (to spare you from typing out responses).

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+  private val RESIZED_IMAGE_COL = "__sparkdl_imageResized"
+  private val INPUT_BUFFER_COL = "__sparkdl_imageBuffer"
+
+  private def validatSchema(schema: StructType) {
```
Nit: `validatSchema` -> `validateSchema`. Otherwise this LGTM!

sueann commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+  private val RESIZED_IMAGE_COL = "__sparkdl_imageResized"
+  private val INPUT_BUFFER_COL = "__sparkdl_imageBuffer"
+
+  private def validateSchema(schema: StructType) {
```
just to confirm, this runs on the Dataset sent into transform right

MrBago replied:
Yep. It's also used in `transformSchema` which, my understanding is, is used by Pipelines to ensure that the pipeline is valid (maybe during a call to pipeline.transform?).

thunterdb replied:
Yes, this is because the pipeline API tries to be more lazy than what Spark allows, especially when combining estimators into a pipeline.

Nit by the way: `: Unit = {` As written, it will not be compatible with Scala 3.

MrBago replied:
Scala 3 requires all methods to have return types?

sueann commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+      Vectors.dense(features.toArray)
+    }
+
+    DebugRowOps
```
why is this called `Debug...`? xD @thunterdb

MrBago replied:
Yes @thunterdb, why is it named Debug... :).

thunterdb replied:
Well, it is called debug, because it was written with idea that it would be soon replaced by `OptimizedRowOps`, but got optimized enough in the process...

You are not supposed to access this internal function, but at the same time it seems useful to expose it.

sueann commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+
+  test("Test test_net on a known data sample.") {
```
do you need `"Test named image featurizer runs on our image samples"` given this test?

MrBago replied:
The "out image samples" test is more of an integration test, it ensures that our featurizer works with a data frame created by "readImages" and makes sure that in can handle images of different sizes and shapes. I can take it out if you think it's too much.

sueann replied:
ah ok. maybe rename it to "Test named image featurizer runs on image dataframes" ?

sueann reviewed (commented):
just a few questions for you @MrBago and @thunterdb. would be good for tim to check the tensorframes part but i'm okay with this as long as the tests run fine.

sueann reviewed (approved):
lgtm. just one small comment. thanks!

thunterdb commented on python/sparkdl/transformers/keras_applications.py:
```diff
@@ -141,5 +170,7 @@ def _imagenet_preprocess_input(x, input_shape):
     "InceptionV3": InceptionV3Model,
     "Xception": XceptionModel,
     "ResNet50": ResNet50Model,
+    "VGG16": VGG16Model,
```
We should merge that upstream, separately. This is a comment for the future, no need to act on it.

MrBago replied:
Sorry that was my mistake, it only showed up on this PR because i failed to push the image-utils branch last night :(.

thunterdb commented on python/tests/transformers/named_image_test.py:
```diff
@@ -60,6 +60,8 @@ class NamedImageTransformerBaseTestCase(SparkDLTestCase):
     __test__ = False
     name = None
+    # Allow subclasses to force number of partitions - a hack to avoid OOM issues
+    numPartitionsOverride = None
```
We should address this issue in tensorframes too, this is annoying.

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+  final val outputCol: Param[String] = new Param[String](this, "outputCol", "output column name")
+
+  // TODO: make modelName a param
+  val modelName = "_test"
```
do we need this to be public?

MrBago replied:
this is just a placeholder for the param. I've fixed it #2 .

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+    val shapeHints = new ShapeDescription(
+      /*
+      // TODO: map doesn't seem to matter
+      out = Map("input" -> Shape(299 * 299 * 3), "output" -> Shape(1234)),
```
indeed, it can infer it from the graph.

MrBago replied:
Is there any benefit to me trying to put the right thing here?

MrBago replied:
Tim what should out be here? Is it ok if I leave it as empty?

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
```diff
@@ -0,0 +1,198 @@
+  /**
+   * A set of supported models that users can access by name.
+   */
+  val supportedModels: Set[String] = _supportedModels.flatMap { case (key, _) =>
```
`.filter` instead

thunterdb commented on .travis.yml:
```diff
@@ -27,7 +27,7 @@ install:
   - ./build/sbt -Dspark.version=$SPARK_VERSION -Dscala.version=$SCALA_VERSION "set test in assembly := {}" assembly
   - ./build/sbt -Dspark.version=$SPARK_VERSION -Dscala.version=$SCALA_VERSION coverage test coverageReport
-  - SPARK_HOME=$HOME/.cache/spark-versions/$SPARK_BUILD ./python/run-tests.sh
+  - SPARK_HOME=$HOME/.cache/spark-versions/$SPARK_BUILD RUN_ONLY_LIGHT_TESTS=True ./python/run-tests.sh
```
this is not used in this PR. I really want to refactor `run-tests.sh` into a python script with a real interface eventually.

thunterdb commented:
@MrBago some tiny comments. Otherwise I am looking forward to it.

MrBago commented:
@thunterdb I think I've addressed your comments, take a look when you get a chance.

btw the plan is to merge this "down", meaning we should start merging with #2 .

thunterdb commented:
@MrBago LGTM
